datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -783,7 +783,6 @@ pub async fn from_substrait_rel(
 
                 let t = ctx.table(table_reference.clone()).await?;
                 let t = ensure_schema_compatability(t, substrait_schema)?;
-                let t = t.into_optimized_plan()?;
                 extract_projection(t, &read.projection)
```

Comment:
I believe with these changes the `extract_projection` may fail (or screw up) the plan if it is present on a case where ensure_schema_compatibility also adds a project into the table, since it'll override it: https://github.com/apache/datafusion/blob/60623afe6e2f44e1312c03b0eb1b06f622097f65/datafusion/substrait/src/logical_plan/consumer.rs#L344. (That's I think why the ensure_schema_compatibility originally had the extra projects and optimize call.) So we'll need to make the `extract_projection` a bit smarter so that it handles a project-of-a-project, I think.

We should also add a test for that case if possible, since I don't think we have any atm..

Reply:
I don't think this is true, actually. Even before the changes, an optimize call would effectively do the same thing, it would push projections into TableScan that might have been overridden by extract_projection call afterwards.

it's actually safe to override them though, because projections provided in the substrait ReadRel are guaranteed to be the subset of the base_schema, right? In other words if ReadRel contains projections the number of clumns in base_schema no longer matters anyway.

P.S. Now that I think of it, we can actually get away with dumbing `extract_projection` down instead because now it will never be passed a `Projection` node, only ever a `TableScan`, so we no longer need to handle Projection case in the match.

Reply:
>  I don't think this is true, actually. Even before the changes, an optimize call would effectively do the same thing, it would push projections into TableScan that might have been overridden by extract_projection call afterwards.

Yeah, possible that it was buggy 😅 the idea I think was that the ensure_schema_compat would leave the Projection node there, and then extract_projection would trigger the second path rather than override the scan.project

> it's actually safe to override them though, because projections provided in the substrait ReadRel are guaranteed to be the subset of the base_schema, right?

Not quite, since the order / indexing might be different. I.e. if 
- the dataset has columns A, B, C, D (in that order), and 
- Substrait ReadRel's base_schema has C, B, A (in that order) + a project [0, 1]
then we should end up with a scan with a project of [3, 2], I think.

> P.S. Now that I think of it, we can actually get away with dumbing extract_projection down instead because now it will never be passed a Projection node, only ever a TableScan, so we no longer need to handle Projection case in the match.

Yup, that sounds right and a nice improvement!

Reply:
> Not quite, since the order / indexing might be different. I.e. if

yup, you're right. projected_schema will probably be fine, but projections themselves can be mixed up.

Reply:
> @Blizzara I'm thinking of switching gears a bit and curious what you think. I think the easiest solution here will be to first "apply" substrait projections to substrait base_schema and only then use `ensure_schema_compatability` with the resulting trimmed down schema as an input.

I think that'd be fine by me!

> One thing I'm unsure about is what to do when data types between table schema and a substrait base_schema do not match for a field which is projected away by substrait anyway. For example
> 
> * table has schema (a int, b str)
> * substrait base_schema is (a int, b int) # note that b has an incorrect type
> * substrait projections field is [0]
> 
> If I trim down base_schema first and only then check compatibility with the actual table, this won't throw an error, it will be as if column b was never mentioned in the plan at all. wdyt? Does that sound like the "right" behavior?

Also fine by me, given as long as the plan can safely execute (all columns used in any way exist and are of right type) then I don't see a need to fail that. But one could as well argue the whole included schema should be validated for consistency reasons, so dunno if @westonpace @vbarua have stricter opinions here?

> (On a related note, I think the same needs to applied to LocalFiles ReadType as well, which for some reason right now outright ignores base_schema. I think if we allow NamedTable base_schema to be different from the actual table schema, makes sense to do the same for LocalFiles).

Yeah, it wasn't done yet just to keep the PR smaller, but agreed the behavior should be same across all source types.

Reply:
@Blizzara fixed with the approach outlined above. btw, turns out there was a [test](https://github.com/apache/datafusion/blob/c412c74a7e6874ee09b1749aec6fd2a0bc72faed/datafusion/substrait/tests/cases/substrait_validations.rs#L120) already that caught this, but apparently I did a classic rookie move in my initial commit and fixed the test instead of fixing the code :laughing: